rust/arrow/src/array/array_string.rs

```diff
@@ -126,25 +126,28 @@ impl<OffsetSize: StringOffsetSizeTrait> GenericStringArray<OffsetSize> {
     }
 
     pub(crate) fn from_vec(v: Vec<&str>) -> Self {
-        let mut offsets = Vec::with_capacity(v.len() + 1);
-        let mut values = Vec::new();
+        let mut offsets =
+            MutableBuffer::new((v.len() + 1) * std::mem::size_of::<OffsetSize>());
+        let mut values = MutableBuffer::new(0);
```

Comment:
Might be interesting to test here whether it is cheaper to calculate offsets + total size first for `values` buffer based on final `length_so_far` so that it doesn't require extra allocations.

Reply:
So something like

```rust
for s in &v {
            length_so_far = length_so_far + OffsetSize::from_usize(s.len()).unwrap();
            offsets.extend_from_slice(length_so_far.to_byte_slice());
        }
        
let mut values = MutableBuffer::new(length_so_far);
for s in &v {
    values.extend_from_slice(s.as_bytes());
}
   ```

Reply:
Probably the offset buffer is also faster after converting it to `typed_data_mut` and iterating it with `iter_mut`, this way it can skip the capacity checks.

Reply:
Both great ideas. If it is ok for you, I will leave them to future work, as I am focusing my work on the `MutableBuffer` atm.

Reply:
I.e. the general direction that I am pushing towards is to stop using `Vec` and replace it by `MutableBuffer` to avoid extra allocations. Once that is in place, we can replace these by `iter_mut`.

> also current conversion to a byte slice may add some overhead?

I believe so, as `to_byte_slice` returns a slice of unknown size, both for `&T` and `&[T]`. Either the compiler optimizes it, or there is an extra cost. I benched a 10% diff in building buffers when introduced a method that was not doing bound checks on these. IMO we should be using `to_le_bytes`, and have a method `MutableBuffer::push<ToByteSlice>`. I think we need the crate `byteorder` because AFAIK std's `to_le_bytes` does not have a trait (which we need).